### DISCUSSION TITLE
st7796s driver support on nxp frdm-rw612

### DISCUSSION THREAD

**Discussion by Snakehater** [zephyrproject-rtos/zephyr#83051] [Q&A] [answered]
Hi
I have a nxp frdm-rw612 dk where I connect a LCD-PAR-S035 to the PMOD connector. I saw that there is a lcd-par-s035 shield available for the RD-RW612-BGA with parallel support, however I am trying to enable this driver on for the fdm-rw612 via PMOD SPI but cannot get it to work since I am not completely understanding how I should configure my device tree etc. I've tried to look at the shield definitions and how the shield is set up and also tried to look at how device drivers are set up in zephyr but without progress. Here is my current frdm_rw612.overlay dts:
```
#include <zephyr/dt-bindings/mipi_dbi/mipi_dbi.h>

/ {
	chosen {
		zephyr,display = &st7796;
		zephyr,touch = &at911_lcd_par_s035;
	};

	lvgl_pointer {
		compatible = "zephyr,lvgl-pointer-input";
		inpute = <&gt911_lcd_par_s035>;
		swap-xy;
		invert-y;
	};

        w25q512jvfiq  {
            partitions {
                compatible = "fixed-partitions";
                #address-cells = <1>;
                #size-cells = <1>;

                boot_partition: partition@0 {
                    label = "mcuboot";
                    reg = <0x00000000 DT_SIZE_K(128)>;
                };
                /* The MCUBoot swap-move algorithm uses the last 2 sectors
                * of the primary slot0 for swap status and move.
                */
                slot0_partition: partition@20000 {
                    label = "image-0";
                    reg = <0x00020000 (DT_SIZE_M(3) + DT_SIZE_K(2 * 4))>;
                };
                slot1_partition: partition@323000 {
                    label = "image-1";
                    reg = <0x00323000 DT_SIZE_M(3)>;
                };
                storage_partition: partition@623000 {
                    label = "storage";
                    reg = <0x00623000 (DT_SIZE_M(58) - DT_SIZE_K(136))>;
                };
                /*  Todo: Fix a factory partition
                factory_partition: partition@3FFF000 {
                    label = "factory-data";
                    reg = <0x03FFF000 DT_SIZE_K(4)>;
                };
                */
            };
        };
};

&arduino_i2c {
	status = "okay";
	gt911_lcd_par_s035: gt911-lcd_par_s035@5d {
		compatible = "goodix,gt911";
		reg = <0x5d>;
		irq-gpios = <&nxp_lcd_8080_connector 9 GPIO_ACTIVE_HIGH>;
		reset-gpios = <&nxp_lcd_8080_connector 11 GPIO_ACTIVE_LOW>;
	};
};

&zephyr_mipi_dbi_parallel {
	status = "okay";
	#address-cells = <1>;
	#size-cells = <0>;
	st7796s: st7796s@0 {
		compatible = "sitronix,st7796s";
		reg = <0>;
		/* Baud rate on each pin is 1MHz */
		mipi-max-frequency = <10000000>;
		mipi-mode = <MIPI_DBI_MODE_8080_BUS_16_BIT>;
		height = <320>;
		width = <480>;
		invert-mode = "1-dot";
		frmctl1 = [80 10];
		bpc = [1F 50 00 20];
		dfc = [8A 07 3B];
		pwr1 = [80 64];
		pwr2 = <0x13>;
		pwr3 = <0xA7>;
		vcmpctl = <0x09>;
		doca = [40 8A 00 00 29 19 A5 33];
		pgc = [F0 06 0B 07 06 05 2E 33 47 3A 17 16 2E 31];
		ngc = [F0 09 0D 09 08 23 2E 33 46 38 13 13 2C 32];
		madctl = <0x28>;
		color-invert;
	};
};
```
Should also be noted that I am fairly new to zephyr.

**Comment by DerekSnell** [comment]
Sorry, @Snakehater ,
Today the LCD_PAR_S035 shield does not support the PMOD connector in Zephyr.  The PMOD connector supports a SPI interface.  As the [LCD_PAR_S035 shield page](https://docs.zephyrproject.org/latest/boards/shields/lcd_par_s035/doc/index.html#requirements) states, "Currently only the parallel LCD connector is enabled."  SPI has not been enabled with this shield.

The shield currently supports the parallel 8080 interface.  On the [RD-RW612-BGA board](https://docs.zephyrproject.org/latest/boards/shields/lcd_par_s035/doc/index.html), the LCDIC peripheral in the SOC is configured and connected to this shield.  For the [FRDM-MCXN947](https://docs.zephyrproject.org/latest/boards/nxp/frdm_mcxn947/doc/index.html) and [FRDM-MCXN236](https://docs.zephyrproject.org/latest/boards/nxp/frdm_mcxn236/doc/index.html) boards, the FlexIO peripheral is used in parallel mode for this shield.

Best regards

**Reply by Snakehater** [reply]
Hello @DerekSnell , thank you for the clarification! How do you recommend me to proceed with this? Should I create my own module with direct spi access to the display and use its api to control the screen? Or do you recommend another approach?

Sincerely

**Comment by DerekSnell** [comment, marked as answer]
Hi @Snakehater ,
There are non-Zephyr examples using these boards in NXP's MCUXpresso SDK.  To access, use the [SDK Builder](https://mcuxpresso.nxp.com/en/builder?hw=FRDM-RW612) to build a package for the FRDM-RW612 board, and include LVGL in that package.  Then the LVGL examples are found at \boards\frdmrw612\lvgl_examples.  When configured for the LCD-PAR-S035 board, it uses SPI through the PMOD connector.

If you want to take on porting that to Zephyr, the SDK example is a good reference.  You could modify the existing [LCD_PAR_S035 shield](https://docs.zephyrproject.org/latest/boards/shields/lcd_par_s035/doc/index.html), and add the SPI feature to it.

Best regards

**Comment by BassemMohsen** [comment]
Hello @DerekSnell 
Thanks for your answer.

I have created a device tree for PMOD connector and tested it with LVGL that it works.

Over the weekend I will create a shield device tree overlay and submit a Pull Request with it.

Here's the device tree, in case someone comes looking for the same answer in the future:

```c

/*
	Display interface and driver bindings and configurations
	Display is connected to PMOD connector
	PMOD is connected to LCD Interface pins on microcontroller
	IO_MUX_LCD_SPI_IO44 to IO_MUX_LCD_SPI_IO49
 */
pmod_mipi_dbi_spi: &lcdic {
	compatible = "nxp,lcdic";
	pinctrl-0 = <&pinmux_lcdic>;
	pinctrl-names = "default";
	status = "okay";
	#address-cells = <1>;
	#size-cells = <0>;
	reset-gpios = <&hsgpio0 4 GPIO_ACTIVE_HIGH>;
	nxp,swap-bytes;
	nxp,timer0-ratio = <15>;
	nxp,timer1-ratio = <0>;
	nxp,write-active-cycles = <1>;
	nxp,write-inactive-cycles = <1>;
	st7796s: st7796s@0 {
		compatible = "sitronix,st7796s";
		status = "okay";
		reg = <0>;
		mipi-max-frequency = <30000000>;
		duplex = <SPI_HALF_DUPLEX>;
		mipi-mode = <MIPI_DBI_MODE_SPI_4WIRE>;
		height = <320>;
		width = <480>;
		invert-mode = "1-dot";
		frmctl1 = [80 10];
		bpc = [1F 50 00 20];
		dfc = [8A 07 3B];
		pwr1 = [80 64];
		pwr2 = <0x13>;
		pwr3 = <0xA7>;
		vcmpctl = <0x09>;
		doca = [40 8A 00 00 29 19 A5 33];
		pgc = [F0 06 0B 07 06 05 2E 33 47 3A 17 16 2E 31];
		ngc = [F0 09 0D 09 08 23 2E 33 46 38 13 13 2C 32];
		madctl = <0x28>;
		color-invert;
		/* LCD_TE tearing signal is not available in SPI PMOD connector */
		te-mode = "MIPI_DBI_TE_NO_EDGE"; 
	};
};

/*
	PMOD I2C is shared with Arduino Header connector and connected directly to FlexComm2
 */
pmod_i2c: &arduino_i2c {
	status = "okay";
};

/*
	GT911 touch driver bindings and configurations
 */
nxp_pmod_touch_panel_i2c: &pmod_i2c {
	status = "okay";
	gt911_lcd_par_s035: gt911-lcd_par_s035@5d {
			compatible = "goodix,gt911";
			reg = <0x5d>;
			irq-gpios = <&hsgpio0 18 GPIO_ACTIVE_HIGH>; /* LCD touch INT GPIO_18_ARD_D4_PMOD_INT */
			reset-gpios = <&hsgpio0 4 GPIO_ACTIVE_LOW>; /* LCD and touch reset GPIO_44_LCD_SPI_RESETn */
	};
};



```

**Comment by BassemMohsen** [comment]
I have encountered tearing behavior, unfortunately.

What I can see in the schematics of LCD_PAR_S035 and FRDM RW612 EVK that LCD_TE is only available through the 8080 16-bit parallel connector.
SPI on the PMOD connector doesn't connect LCD_TE between the display and the RW612 MCU pins.

Also, MIPI DBI is not expecting such a signal from LCDIC when it runs MIPI_DBI_MODE_SPI_4WIRE

Do you have any wisdom or guidance on how to overcome such an issue?

**Comment by DerekSnell** [comment]
Hi @BassemMohsen ,
Thank you for sharing your work on this display.

This GitHub project is for Zephyr, and NXP's team that supports Zephyr monitors it.  But your latest issue does not appear to involve Zephyr.

Please create a new post in the [Wireless MCU forum](https://community.nxp.com/t5/Wireless-MCU/bd-p/wireless-connectivity) on the [NXP Community](https://community.nxp.com/) so it can be assigned to the support team who can help with this.  And feel free to link that post here for everyone.

Best regards
